File: Python/DPR.py
from prettytable import PrettyTable # Import library untuk menampilkan data dalam bentuk tabel

class DPR: # Membuat class DPR
    __name = "" # Membuat atribut __name berjenis private
    __idNumber = 0 # Membuat atribut __idNumber berjenis private
    __bidang = "" # Membuat atribut __bidang berjenis private
    __partai = "" # Membuat atribut __partai berjenis private

    def __init__(self, name = "", idNumber = 0, bidang = "", partai = ""): # Membuat constructor
        self.__name = name # Menginisialisasi atribut __name
        self.__idNumber = idNumber # Menginisialisasi atribut __idNumber
        self.__bidang = bidang # Menginisialisasi atribut __bidang
        self.__partai = partai # Menginisialisasi atribut __partai

    def setName(self, name): # Setter untuk atribut __name
        self.__name = name # Menginisialisasi atribut __name
    
    def setIdNumber(self, idNumber): # Setter untuk atribut __idNumber
        self.__idNumber = idNumber # Menginisialisasi atribut __idNumber

    def setBidang(self, bidang): # Setter untuk atribut __bidang
        self.__bidang = bidang # Menginisialisasi atribut __bidang

    def setPartai(self, partai): # Setter untuk atribut __partai
        self.__partai = partai # Menginisialisasi atribut __partai

    def getName(self): # Getter untuk atribut __name
        return self.__name # Mengembalikan nilai atribut __name

    def getIdNumber(self): # Getter untuk atribut __idNumber
        return self.__idNumber # Mengembalikan nilai atribut __idNumber

    def getBidang(self): # Getter untuk atribut __bidang
        return self.__bidang # Mengembalikan nilai atribut __bidang

    def getPartai(self): # Getter untuk atribut __partai
        return self.__partai # Mengembalikan nilai atribut __partai

    def commandOptions(self): # Membuat method untuk menampilkan menu command
        print("==============================")
        print("|        CRUD Program        |") # CRUD = Create, Read, Update, Delete
        print("==============================")
        print("|  0. Exit the program       |") # 0 untuk keluar dari program
        print("|  1. Add DPR data           |") # 1 untuk menambahkan data mahasiswa
        print("|  2. Update DPR data        |") # 2 untuk mengubah data mahasiswa
        print("|  3. Delete DPR data        |") # 3 untuk menghapus data mahasiswa
        print("|  4. Show list of DPR       |") # 4 untuk menampilkan semua data dalam daftar mahasiswa
        print("==============================")
        print("Enter your choice : ", end = '')

    def addDPRData(self, dprList, temp, name, idNumber, bidang, partai): # Membuat method untuk menambahkan data mahasiswa
    # Proses setter data inputan ke setiap atribut dari objek temp
        temp.setName(name)
        temp.setIdNumber(idNumber)
        temp.setBidang(bidang)
        temp.setPartai(partai)

        if not dprList: # Jika list kosong, langsung masukan data ke dalam list
            dprList.append(temp) # Memasukan data ke dalam list
            print("\n[SUCCESS] Data has been added.\n\n") # Menampilkan pesan bahwa data berhasil dimasukan ke dalam list
        else: # Jika list tidak kosong, lakukan pengecekan terlebih dahulu
            find = False # Untuk mengecek apakah data sudah ada atau belum, nilai defaultnya adalah false yang menandakan data belum ada

            i = 0 # Untuk mengecek data satu persatu
            while not find and i < len(dprList):
                if temp.getIdNumber() == dprList[i].getIdNumber(): # Jika NIM sudah ada, maka data tidak akan dimasukan ke dalam list dan akan menampilkan pesan error
                    find = True # Jika NIM sudah ada, maka nilai find akan berubah menjadi true
                    print("\n[ERROR] ID number already exists, please add data with a different ID number.\n\n") # Pesan error
                i += 1 # Iterasi untuk mengecek data selanjutnya

            if not find : # Jika data tidak ditemukan, maka data akan dimasukan ke dalam list
                dprList.append(temp) # Memasukan data ke dalam list
                print("\n[SUCCESS] Data has been added.\n\n") # Menampilkan pesan bahwa data berhasil dimasukan ke dalam list

    def updateDPRData(self, dprList, idNumber): # Membuat method untuk mengubah data mahasiswa
        find = False # Untuk mengecek apakah data ada atau tidak, nilai defaultnya adalah false yang menandakan data tidak ada
        i = 0 # Membuat iterator untuk mengecek data dalam list

        while not find and i < len(dprList): # do while untuk mengecek apakah data yang akan diubah ada di dalam dprList atau tidak
            if dprList[i].getIdNumber() == idNumber: # Jika data yang akan diubah ditemukan di dalam dprList, maka akan meminta beberapa masukan
                find = True # Jika data ditemukan, maka nilai find akan berubah menjadi true
                self.__name = input("\nEnter Nama         : ") # Meminta masukan untuk mengubah nama
                self.__bidang = input("Enter bidang     : ") # Meminta masukan untuk mengubah jurusan
                self.__partai = input("Enter partai     : ") # Meminta masukan untuk mengubah fakultas

                dprList[i] = DPR(self.__name, idNumber, self.__bidang, self.__partai) # Mengubah data yang sudah ada di dalam dprList dengan data baru
                print("\n[SUCCESS] Data has been changed.\n\n") # Menampilkan pesan bahwa data berhasil diubah
            i += 1 # Iterasi untuk mengecek data selanjutnya

        if not find: # Jika data tidak ditemukan, maka akan menampilkan pesan error
            print("\n[ERROR] DPR data not found in the list.\n\n") # Pesan error

    def deleteDPRData(self, dprList, idNumber): # Membuat method untuk menghapus data mahasiswa
        find = False # Untuk mengecek apakah data ada atau tidak, nilai defaultnya adalah false yang menandakan data tidak ada
        i = 0 # Membuat iterator untuk mengecek data dalam list

        while not find and i < len(dprList): # do while untuk mengecek apakah data yang akan diubah ada di dalam dprList atau tidak
            if dprList[i].getIdNumber() == idNumber: # Jika data yang akan dihapus ditemukan di dalam dprList, maka data akan dihapus
                find = True # Jika data ditemukan, maka nilai find akan berubah menjadi true
                del dprList[i] # Menghapus data yang ditemukan

                print("\n[SUCCESS] Data has been deleted.\n\n") # Menampilkan pesan bahwa data berhasil dihapus  
            i += 1 # Iterasi untuk mengecek data selanjutnya

        if not find: # Jika data tidak ditemukan, maka akan menampilkan pesan error
            print("\n[ERROR] ID number not exist!\n\n") # Pesan error


    def showListOfDPR(self, dprList): # Membuat method untuk menampilkan daftar mahasiswa
        if not dprList: # Jika list kosong, maka akan menampilkan pesan error
            print("\n[WARNING] The DPR list is still empty.\n\n") # Pesan error
        else: # Jika list tidak kosong, maka akan menampilkan data mahasiswa
            print() # Untuk memberikan jarak antara menu command dan daftar mahasiswa
        
            table = PrettyTable() # Membuat objek table dari class PrettyTable
            table.field_names = ["ID", "Nama", "bidang", "partai"] # Membuat header untuk tabel

            for i, DPR in enumerate(dprList): # Looping untuk menampilkan data mahasiswa
                table.add_row([DPR.getIdNumber(), DPR.getName(), DPR.getBidang(), DPR.getPartai()]) # Menambahkan data mahasiswa ke dalam tabel

            print(table) # Menampilkan tabel
            print('\n') # Menambah baris kosong

File: PHP/DPR.php
<?php

class DPR // Membuat class DPR
{
    /* Membuat atribut berjenis private */
    private $id; // Untuk menyimpan id
    private $nama; // Untuk menyimpan nama
    private $bidang; // Untuk menyimpan bidang
    private $partai; // Untuk menyimpan partai
    private $foto; // untuk menyimpan foto

    /* Konstruktor */
    public function __construct($id = '', $nama = 0, $bidang = '', $partai = '', $foto = '') // Membuat konstruktor dengan isian dari parameter
    {
        $this->id = $id;         // Meng-assign atribut id dengan nilai dari parameter id
        $this->nama = $nama; // Meng-assign atribut nama dengan nilai dari parameter nama
        $this->bidang = $bidang;       // Meng-assign atribut bidang dengan nilai dari parameter bidang
        $this->partai = $partai;   // Meng-assign atribut partai dengan nilai dari parameter partai
        $this->foto = $foto; // meng-asign atribut foto dengan nilai dari parameter foto
    }

    /* Setter dan Getter */
    public function setId($id) // Setter untuk atribut id
    {
        $this->id = $id; // Meng-assign atribut id dengan nilai dari parameter id
    }

    public function setNama($nama) // Setter untuk atribut nama
    {
        $this->nama = $nama; // Meng-assign atribut nama dengan nilai dari parameter nama
    }

    public function setBidang($bidang) // Setter untuk atribut bidang
    {
        $this->bidang = $bidang; // Meng-assign atribut bidang dengan nilai dari parameter bidang
    }

    public function setPartai($partai) // Setter untuk atribut partai
    {
        $this->partai = $partai; // Meng-assign atribut partai dengan nilai dari parameter partai
    }

    public function setFoto($foto) // Setter untuk atribut foto
    {
        $this->foto = $foto; // Meng-assign atribut foto dengan nilai dari parameter foto
    }

    public function getId() // Getter untuk atribut id
    {
        return $this->id; // Me-return nilai dari atribut id
    }

    public function getNama() // Getter untuk atribut nama
    {
        return $this->nama; // Me-return nilai dari atribut nama
    }

    public function getBidang() // Getter untuk atribut bidang
    {
        return $this->bidang; // Me-return nilai dari atribut bidang
    }

    public function getPartai() // Getter untuk atribut partai
    {
        return $this->partai; // Me-return nilai dari atribut partai
    }

    public function getFoto() // Getter untuk atribut foto
    {
        return $this->foto; // Me-return nilai dari atribut foto
    }

    /* Method untuk CRUD */
    public function addDPRData(&$DPRlist, $temp) // method untuk menambah data anggota dpr
    {
        if (empty($DPRlist)) { //Jika array kosong
            array_push($DPRlist, $temp); // Menambahkan data anggota dpr ke array
            echo "Data Anda Berhasil Ditambahkan" . "<br>"; // Menampilkan pesan bahwa  
        } else { // Jika array tidak kosong
            $find = false; // Untuk mengecek apakah data sudah ada atau belum, nilai defaultnya false yang menandakan data belum ada
            $i = 0; // Membuat iterator untuk mengecek data dalam list

            do { // do while untuk mengecek apakah data sudah ada atau belum, pengecekkan menggunakan nama sebagai primary key
                if ($temp->getNama() == $DPRlist[$i]->getNama()) { //Jika nama sudah ada, maka data tidak akan ditambahkan ke dalam array
                    $find = true; // Mengubah nilai find menjadi true
                    echo "Data Nama Telah Digunakan, Silahkan Menambahkan Data Baru dengan Nama Yang Berbeda." . "<br>"; // Menampilkan pesan error
                }
                $i++; // Iterasi untuk mengecek data selanjutnya
            } while ($find == false && $i < count($DPRlist)); // Jika data belum ditemukan dan iterator belum sampai ke akhir array, maka akan melakukan proses pengecekan berikutnya

            if ($find == false) { // Jika data tidak ditemukan, maka     data akan dimasukan ke dalam array
                array_push($DPRlist, $temp); // Memasukan data ke dalam array
                echo "Data Anda Berhasil Ditambahkan" . "<br>"; // Menampilkan pesan bahwa data berhasil dimasukan ke dalam array
            }
        }
    }

    public function updateDPRData(&$DPRlist, $temp, $id, $nama, $bidang, $partai) // Method untuk mengubah data anggota dpr
    {
        // Proses setter data inputan ke setiap atribut dari objek temp  
        $temp->setId($id);
        $temp->setNama($nama);
        $temp->setBidang($bidang);
        $temp->setPartai($partai);

        $find = false; // Untuk mengecek apakah data ada di dalam array atau tidak, nilai defaultnya false yang menandakan data tidak ada
        $i = 0; // Membuat iterator untuk mengecek data dalam list

        do { // do while untuk mengecek apakah data yang akan diubah ada di dalam array, pengecekan menggunakan nama sebagai primary key
            if ($temp->getNama() == $DPRlist[$i]->getNama()) { // Jika data anggota dpr ditemukan, maka akan langsung diubah datanya 
                $find = true; // Mengubah nilai find menjdai true
                $DPRlist[$i] = $temp; // Ubah data anggota dpr di array
                echo "Data Anda Berhasil Diubah" . "<br>"; // Menampilkan pesan berhasil mengubah data
            }
            $i++; // Iterasi untuk mengecek data selanjutnya
        } while ($find == false && $i < count($DPRlist)); // Jika data belum ditemukan dan iterator belum sampai ke akhir array, maka akan melakukan proses pengecekan berikutnya

        if ($find == false) { // Jika data tidak ditemukan, maka akan menampilkan pesan error
            echo "Gagal Mengubah Data Tidak Ada Didalam List" . "<br>";
        }
    }

    public function deleteDPRData(&$DPRlist, $nama) // Method untuk menghapus data anggota dpr
    {
        $find = false; // Untuk mengecek apakah data ada di dalam array atau tidak, nilai defaultnya false yang menandakan data tidak ada
        $i = 0; // Membuat iterator untuk mengecek data dalam list

        do { // do while untuk mengecek apakah data yang akan dihapus ada di dalam array, pengecekan menggunakan nama sebagai primary key
            if ($nama == $DPRlist[$i]->getNama()) { // Jika data anggota dpr ditemukan, maka akan langsung dihapus datanya
                $find = true; // Mengubah nilai find menjdai true
                array_splice($DPRlist, $i, 1); // Menghapus data anggota dpr di dalam array
                echo "Data Anda Berhasil Dihapus" . "<br>"; // Menampilkan pesan berhasil menghapus data
            }
            $i++; // Iterasi untuk mengecek data selanjutnya
        } while ($find == false && $i < count($DPRlist)); // Jika data belum ditemukan dan iterator belum sampai ke akhir array, maka akan melakukan proses pengecekan berikutnya

        if ($find == false) { // Jika data tidak ditemukan, maka akan menampilkan pesan error
            echo "Gagal Menghapus Data Nama Tidak Ada didalam List!" . "<br>";
        }
    }

    public function showListOfDPR($DPRlist) // Method untuk menampilkan semua data anggota dpr di dalam array
    {
        if (empty($DPRlist)) { // Jika array kosong
            echo "[WARNING] The DPR list is still empty." . "<br>"; // Menampilkan pesan error
        } else { // Jika array ada isinya
            echo "<h3>Data Dari List DPR</h3>"; // Judul Tabel
            echo "<table border='1' cellpadding='5' cellspacing = '0'>";
            echo "<tr>";
            echo "<th>id</th>" . "<th>Nama</th>" . "<th>bidang</th>" . "<th>partai</th>". "<th>foto</th>"; // Header Tabel
            echo "</tr>";
            $no = 1;
            /* Menampilkan semua data anggota dpr */
            for ($i = 0; $i < count($DPRlist); $i++) {
                echo "<tr>";
                // echo "<td>" . $no . '. ' . "</td>";
                echo "<td>" . $DPRlist[$i]->getId() . "</td>";
                echo "<td>" . $DPRlist[$i]->getNama() . "</td>";
                echo "<td>" . $DPRlist[$i]->getBidang() . "</td>";
                echo "<td>" . $DPRlist[$i]->getPartai() . "</td>";
                echo "<td><img src=\"".  $DPRlist[$i]->getFoto() . "\" width='50px' /></td>";
                echo "</tr>";
            }
            echo "</table>";
            echo "<br>";
        }
    }
}